scrape/scrape_test.go

```diff
@@ -5897,3 +5898,48 @@ func TestScrapeLoopAppendSampleLimitReplaceAllSamples(t *testing.T) {
 	}...)
 	requireEqual(t, want, resApp.resultFloats, "Appended samples not as expected:\n%s", slApp)
 }
+
+func TestScrapeLoopDisableStalenessMarkerInjection(t *testing.T) {
+	var (
+		loopDone = atomic.NewBool(false)
+		appender = &collectResultAppender{}
+		scraper  = &testScraper{}
+		app      = func(_ context.Context) storage.Appender { return appender }
+	)
+
+	ctx, cancelLoop := context.WithCancel(context.Background())
+	sl := newBasicScrapeLoop(t, ctx, scraper, app, 10*time.Millisecond)
+	scraper.scrapeFunc = func(ctx context.Context, w io.Writer) error {
+		if _, err := w.Write([]byte("metric_a 42\n")); err != nil {
+			return err
+		}
+		return ctx.Err()
+	}
+
+	// Start the scrape loop.
+	go func() {
+		sl.run(nil)
+		loopDone.Store(true)
+	}()
+
+	// Wait for some samples to be appended.
+	require.Eventually(t, func() bool {
+		appender.mtx.Lock()
+		defer appender.mtx.Unlock()
+		return len(appender.resultFloats) > 2
+	}, 5*time.Second, 100*time.Millisecond, "Scrape loop didn't append any samples.")
+
+	// Disable end of run staleness markers and stop the loop.
+	sl.disableEndOfRunStalenessMarkers()
+	cancelLoop()
```

Comment:
If you comment out `sl.disabledEndOfRunStalenessMarkers.Store(true)` in scrape.go, the test doesn't fail, because context cancel is too abrupt.

Use stop() instead.

```suggestion
	sl.stop()
```